lighthouse-core/gather/gatherers/source-maps.js

```diff
@@ -0,0 +1,118 @@
+/**
+ * @license Copyright 2019 Google Inc. All Rights Reserved.
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except in compliance with the License. You may obtain a copy of the License at http://www.apache.org/licenses/LICENSE-2.0
+ * Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the specific language governing permissions and limitations under the License.
+ */
+'use strict';
+
+/** @typedef {import('../driver.js')} Driver */
+
+const Gatherer = require('./gatherer.js');
+
+/**
+ * This function is careful not to parse the response as JSON, as it will
+ * just need to be serialized again over the protocol, and source maps can
+ * be huge.
+ *
+ * @param {string} url
+ */
+/* istanbul ignore next */
+async function fetchSourceMap(url) {
+  try {
+    // eslint-disable-next-line no-undef
+    const response = await fetch(url);
+    return response.text();
+  } catch (err) {
+    return {errorMessage: err.toString()};
+  }
+}
+
+/**
+ * @fileoverview Gets JavaScript source maps.
+ */
+class SourceMaps extends Gatherer {
+  constructor() {
+    super();
+    /** @type {LH.Crdp.Debugger.ScriptParsedEvent[]} */
+    this._scriptParsedEvents = [];
+    this.onScriptParsed = this.onScriptParsed.bind(this);
+  }
+
+  /**
+   * @param {Driver} driver
+   * @param {string} sourceMapUrl
+   * @return {Promise<{map: LH.Artifacts.RawSourceMap} | {errorMessage: string}>}
+   */
+  async fetchSourceMapInPage(driver, sourceMapUrl) {
+    // TODO: change default protocol timeout?
+    // driver.setNextProtocolTimeout(250);
+    /** @type {string|{errorMessage: string}} */
+    const sourceMapJson =
+      await driver.evaluateAsync(`(${fetchSourceMap})(${JSON.stringify(sourceMapUrl)})`);
+
+    if (typeof sourceMapJson === 'object') {
```

Comment:
this is super sketch :)

what about just making `fetchSourceMap` itself return `{map: LH.Artifacts.RawSourceMap} | {errorMessage: string}`

Reply:
see the comment [here](https://github.com/GoogleChrome/lighthouse/pull/9101/files#diff-7519d8c2231119d4a3feee7c6c8e8cd5R13)

Reply:
> see the comment here

yeah...yeah. Ugh :)

We could maybe follow up with a fetch built in to driver.

Reply:
at least it needs a comment. `if (typeof sourceMapJson === 'object')` doesn't give a clue as to what's actually being tested (that it's an error)

Reply:
it's documented in the comment for `fetchSourceMap`, i've added a comment referencing that + renamed the variable.